Properly shutdown RPC even in the case of `clean_shutdown=False`.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#29148 Properly shutdown RPC even in the case of `clean_shutdown=False`.**

We would skip rpc.join_rpc() in the case of `clean_shutdown=False`.
This would exit the process without properly cleaning up the local RPCAgent
resulting in a crash.

As a result, to fix this we still call rpc.join_rpc() even in an unclean
shutdown. Note that, rpc.join_rpc() needs to be replaced with a local
`shutdown` call eventually since we need a way to shutdown the local RPC agent
properly.

Differential Revision: [D18306941](https://our.internmc.facebook.com/intern/diff/D18306941/)

Commit:
Properly shutdown RPC even in the case of `clean_shutdown=False`.

We would skip rpc.join_rpc() in the case of `clean_shutdown=False`.
This would exit the process without properly cleaning up the local RPCAgent
resulting in a crash.

As a result, to fix this we still call rpc.join_rpc() even in an unclean
shutdown. Note that, rpc.join_rpc() needs to be replaced with a local
`shutdown` call eventually since we need a way to shutdown the local RPC agent
properly.

Differential Revision: [D18306941](https://our.internmc.facebook.com/intern/diff/D18306941/)

[ghstack-poisoned]

rohan-varma commented:
> Note that, rpc.join_rpc() needs to be replaced with a local
> `shutdown` call eventually since we need a way to shutdown the local RPC agent
> properly.

This should be unblocked once https://github.com/facebookincubator/gloo/pull/228 is landed.

Commit:
Update on "Properly shutdown RPC even in the case of `clean_shutdown=False`."


We would skip rpc.join_rpc() in the case of `clean_shutdown=False`.
This would exit the process without properly cleaning up the local RPCAgent
resulting in a crash.

As a result, to fix this we still call rpc.join_rpc() even in an unclean
shutdown. Note that, rpc.join_rpc() needs to be replaced with a local
`shutdown` call eventually since we need a way to shutdown the local RPC agent
properly.

Differential Revision: [D18306941](https://our.internmc.facebook.com/intern/diff/D18306941/)

[ghstack-poisoned]

xush6528 commented on test/dist_utils.py:
```diff
@@ -94,34 +94,38 @@ def new_test_method(self, *arg, **kwargs):
+            # since we need to shutdown the RPC agent. If we don't shutdown the
+            # RPC agent, tests would fail since RPC agent threads, locks and
+            # condition variables are not properly terminated.
             rpc.join_rpc()
```
Yes, replace this with the local shutdown call after the shutdown API is ready.

xush6528 commented on test/dist_utils.py:
```diff
@@ -94,34 +94,38 @@ def new_test_method(self, *arg, **kwargs):
+                        fut = rpc.rpc_async(dst_name, set_termination_signal, args=())
+                        futs.append(fut)
+                    for fut in futs:
+                        assert fut.wait() is None, "Sending termination signal failed."
```
And replace this part with the `rpc.all_shutdown()` API to be added.
